packages/rum-core/src/boot/rumPublicApi.ts

```diff
@@ -16,13 +16,13 @@ import {
   display,
 } from '@datadog/browser-core'
 import { ProvidedSource } from '../domain/rumEventsCollection/error/errorCollection'
-import { CommonContext, User, ActionType } from '../rawRumEvent.types'
+import { CommonContext, User, ActionType, RumEventDomainContext } from '../rawRumEvent.types'
 import { RumEvent } from '../rumEvent.types'
 import { startRum } from './rum'
 
 export interface RumUserConfiguration extends UserConfiguration {
   applicationId: string
-  beforeSend?: (event: RumEvent) => void | boolean
+  beforeSend?: (event: RumEvent, context: RumEventDomainContext<any>) => void | boolean
```

Comment:
I implemented your solution. After a few tests, this seems a bit less strict (TS is confused with large type unions), but maybe a bit easier to understand.

Reply:
what about something like:
```ts
export type RawRumEventCollectedData<
  Event = RawRumEvent, 
  Domain = RumEventDomainContext
> = {
  startTime: RelativeTime
  savedCommonContext?: CommonContext
  customerContext?: Context
  rawRumEvent: Event
  domainContext: Domain
}
```

It could allow us to specify the right types when needed and could have a better TS support.

Reply:
I think my initial solution was superior, then: it was the same, but with the domain context automatically inferred from the event type.

Reply:
In the initial approach, DomainContext types were not easy to discover or reuse for customers.

what I had in mind was for customers to be able to import those DomainContext types in their beforeSend implementation, something like:
```
beforeSend(event, domainContext) {
  if (event.type === 'error') {
    const errorContext = domainContext as RumErrorDomainContext
    ...
  } 
}
```

with the proposed approach, we could make types a bit more strict and declarative, like in `resourceCollection`:
```ts
function processRequest(request: RequestCompleteEvent): RawRumEventCollectedData<RawRumResourceEvent, RumFetchDomainContext | RumXhrDomainContext>
```
which would raise a type error with the current implementation

wdyt?

Reply:
We can totally keep both the individual domain interfaces and the inference upon the RUM event. I'll make something.